src/ThriftSQL/Utils/Iterator.php

```diff
@@ -90,22 +104,22 @@ public function valid() {
 
   /**
    * Rewind the Iterator to the first element
-   * @link http://php.net/manual/en/iterator.rewind.php
    * @return void Any returned value is ignored.
-   * @since 5.0.0
    * @throws \ThriftSQL\Exception
    */
   public function rewind() {
-    if ( $this->runCount > 0 && !$this->allowRerun ) {
+    if ( $this->rewindCount > 1 && !$this->allowRerun ) {
       throw new \ThriftSQL\Exception(
         'Iterator rewound, this will cause the ThriftSQL to execute again. ' .
         'Set `' . __CLASS__ . '::allowRerun(true)` to allow this behavior.'
       );
     }
-    $this->runCount++;
 
-    $this->buffer = array();
-    $this->location = 0;
-    $this->thriftSQLQuery = $this->thriftSQL->query( $this->queryStr )->wait();
+    $this->rewindCount++;
+    if ($this->rewindCount > $this->queryCount) {
+      $this->queryCount++;
+      $this->doQuery();
+    }
+    $this->thriftSQLQuery->wait();
```

Comment:
Seems like extracting out this into `doQuery()` and tracking `queryCount` and `rewindCount` separately just makes this method more complex, is this needed?

Reply:
Well ... if let me decide, I would prevent users from calling `rewind` twice (throw an exception at the second time), the `allowRerun` feature should be disabled, it doesn't make sense. If the user knows what he/she does, he/she should run the query explicitly again.